ccm_web/ecosystem.config.js

```diff
@@ -0,0 +1,14 @@
+// https://pm2.keymetrics.io/docs/usage/docker-pm2-nodejs/
+
+module.exports = {
+  apps: {
+    name: 'CCM - NestJS app',
+    script: 'server/src/main.js',
+    exec_mode: 'cluster',
+    instances: 'max',
+    autorestart: true,
+    max_memory_restart: '1G',
```

Comment:
8GBs~ per pod? Maybe bump this to 1.8?

Reply:
Looks like this is "maximum memory amount before PM2 restarts your app" so in here to prevent a process that takes up too much memory. I wonder if this option should just be able to be dynamically set, like in the environment? 

I'd think you can pass some of these args at runtime in the start script to pm2-runtime. I'd probably set this as a default like we're doing with the MyLA start script but make it configurable. 

It would probably be a calculation of like `(90% * [Max Memory])/[# of threads]`.

Reply:
@jonespm, are you thinking this would just be a simple environment variable, like `pm2-runtine ecosystem.config.js --max_memory_restart=$PM2_MEM_LIMIT`, or would we want to do something more fancy to calculate it on the fly? It looks like you can get limits from OpenShift using a pod template and environment variables, but I don't exactly know how we would do that with a DeploymentConfig so it happens with each new pod (https://docs.openshift.com/container-platform/4.1/nodes/containers/nodes-containers-downward-api.html); even then, I don't know if I can be a 100 percent certain how many processes `pm2` will spawn.

We could also do this in the `ecosystem.config.js` file, something like `max_memory_restart: process.env.PM2_MEM_LIMIT || '1G'`. Does command line arg or this seem better to you?

Reply:
Nope, I think we'd just pass it in and have a default like

```
if [ -z "${PM2_MEM_LIMIT}" ]; then
    PM2_MEM_LIMIT=1
fi
```

I think either way is fine, parameter or js. Didn't realize this was a javascript file and not a plain configuration file. That looks good to me.

Reply:
I guess you don't even have to check and set the default in the config if you're setting a default in the javascript, so your suggestion alone seems fine too.

Reply:
There are ways we could calculate it from the start script too like.

```
MEM_TOTAL=$(grep MemTotal /proc/meminfo | awk '{print $2}');NUM_THREADS=$(grep -c processor /proc/cpuinfo)

MEM_PER_PROCESSOR=$(expr ${MEM_TOTAL} / ${NUM_THREADS})
```

This would be in KB, or just divide by the 1000. Not sure if this is *really* necessary though.

Reply:
That's cool, though it seems a tad fragile with the string parsing. I also tried it in the terminal for CCM Dev, and I don't understand the result; it looks like it's 8 GB per thread, which is not what I expected.

Reply:
Okay, I did the minimum here for now: https://github.com/tl-its-umich-edu/canvas-course-manager-next/pull/174/commits/b12c0194a5fa7fc069234aacd33198ed5250cf3f

I think if I understood why we're getting that result, I'd consider using your approach. Something to think about for the future.

Reply:
Yeah, hmm, /proc/meminfo says it has 32GB, rather than 8.

MemTotal:       32393208 kB

It looks like for [Openshift it should look at ](https://stackoverflow.com/a/63047839/3708872)
cat /sys/fs/cgroup/memory/memory.limit_in_bytes

This is even more direct, and the number of threads [is similarly direct.
](https://stackoverflow.com/a/54112066/3708872)

```
MEM_TOTAL=$(cat /sys/fs/cgroup/memory/memory.limit_in_bytes)
NUM_THREADS=$(nproc) 
MEM_PER_PROCESSOR=$(expr ${MEM_TOTAL} / ${NUM_THREADS})
```

Reply:
I probably spent too much time today looking into this. I think something like this would work in OpenShift right now:

```
TOTAL_MEM=$(cat /sys/fs/cgroup/memory/memory.limit_in_bytes)
NUM_PROC=$(nproc)
PM2_MEM_LIMIT_CALC=$(($TOTAL_MEM * 9 / 10 / $NUM_PROC))
```

but the paths seem to be set up differently in a Docker container. The `cat` line returns a ridiculously long number (`9223372036854771712`). I guess I'm also concerned that this path might change. Looks like there is a `cgget` utility library that might help out. However, I'm a little concerned that things might change in OpenShift, and I wonder if simply making it directly configurable would save us time and heartache over a long period.

Reply:
Yup, fine with just leaving it directly configurable. Humans can figure this out and the 1GB default seems like it should always be fine for us.